sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/sns/SnsIO.java

```diff
@@ -0,0 +1,299 @@
+    }
+
+    /**
+     * Specify credential details and region to be used to write to SNS. If you need more
+     * sophisticated credential protocol, then you should look at {@link
+     * Write#withAWSClientsProvider(AwsClientsProvider)}.
+     */
+    public Write withAWSClientsProvider(String awsAccessKey, String awsSecretKey, Regions region) {
+      return withAWSClientsProvider(awsAccessKey, awsSecretKey, region, null);
+    }
+
+    /**
+     * Specify credential details and region to be used to write to SNS. If you need more
+     * sophisticated credential protocol, then you should look at {@link
+     * Write#withAWSClientsProvider(AwsClientsProvider)}.
+     *
+     * <p>The {@code serviceEndpoint} sets an alternative service host. This is useful to execute
+     * the tests with Kinesis service emulator.
+     */
+    public Write withAWSClientsProvider(
+        String awsAccessKey, String awsSecretKey, Regions region, String serviceEndpoint) {
+      return withAWSClientsProvider(
+          new BasicSnsProvider(awsAccessKey, awsSecretKey, region, serviceEndpoint));
+    }
+
+    /** Max delay in seconds for retries. Mandatory field. */
+    public Write withMaxDelay(Duration maxDelay) {
+      return builder().setMaxDelay(maxDelay).build();
+    }
+
+    /** Max number of times to retry before writing to failures. Mandatory field. */
+    public Write withMaxRetries(int maxRetries) {
+      return builder().setMaxRetries(maxRetries).build();
+    }
+
+    /** Initial delay in seconds for retries. Mandatory field. */
+    public Write withRetryDelay(Duration retryDelay) {
+      return builder().setRetryDelay(retryDelay).build();
+    }
+
+    /** Tuple tag to store results. Mandatory field. */
+    public Write withResultOutputTag(TupleTag<PublishResult> results) {
+      return builder().setResultOutputTag(results).build();
+    }
+
+    @Override
+    public PCollectionTuple expand(PCollection<PublishRequest> input) {
+      checkArgument(getTopicName() != null, "withTopicName() is required");
+      final PCollectionTuple tuple =
+          input.apply(
+              ParDo.of(new SnsWriterFn(this))
+                  .withOutputTags(getResultOutputTag(), TupleTagList.empty()));
+      return tuple;
+    }
+
+    private static class SnsWriterFn extends DoFn<PublishRequest, PublishResult> {
+
+      private static final Logger LOG = LoggerFactory.getLogger(SnsWriterFn.class);
+      private static final Counter SNS_WRITE_FAILURES =
+          Metrics.counter(SnsWriterFn.class, "SNS_Write_Failures");
+
+      private final SnsIO.Write spec;
+      private transient AmazonSNS producer;
+      private transient RetryPolicy retryPolicy;
+
+      public SnsWriterFn(SnsIO.Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup() throws Exception {
+
+        //Initialize SnsPublisher
+        producer = spec.getAWSClientsProvider().createSnsPublisher();
+        retryPolicy =
```

Comment:
Beam has already a [Backoff facility](https://github.com/apache/beam/blob/20aff13d36b4ac7d9b9a29fd9f6fb3e80a84214a/sdks/java/core/src/main/java/org/apache/beam/sdk/util/FluentBackoff.java#L30) so probably worth using that one to avoid the extra dependency. Please take a look at the use of [Backoff in SolrIO](https://github.com/apache/beam/blob/20aff13d36b4ac7d9b9a29fd9f6fb3e80a84214a/sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java#L607) to see how it works, it seems like an easy rewrite. Notice that the default case is no retry so probably worth doing the same here, and that the user can parametrize the retry via a function like done [there too](https://github.com/apache/beam/blob/20aff13d36b4ac7d9b9a29fd9f6fb3e80a84214a/sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java#L261)